### PR TITLE
bump tree-sitter in Cargo.toml

### PR DESCRIPTION
**Pull request by tek**
@VixieTSQ I updated the version and ran `cargo update`, is that doing anything for you?

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -19,7 +19,7 @@ include = [
 path = "bindings/rust/lib.rs"
 
 [dependencies]
-tree-sitter = "0.19.4"
+tree-sitter = "0.20.6"
```
Comment (VixieTSQ) [inline Cargo.toml:22]: could you remove the patch number? I think that's causing problems. Cargo will choose the latest patch anyways.
```suggestion
tree-sitter = "0.20"
```

Comment (VixieTSQ) [reply Cargo.toml:22]: NEVERMIND! I was just an idiot and didn't use the correct branch!! gahhh. This is still work changing though. You don't normally specify the branch.

**Comment by VixieTSQ** [comment]
This works! Says the compiler at least. And that's because the compiler *assumes* the `Language` it's getting back from the FFI call is indeed a version 0.20 `Language`. But I'm scared that might not be correct, which would cause UB. And I assume the `package.json` specifies dependencies for the javascript or whatever? I've never used it but it looks like it still specifies 0.19. Whatever you did to generate the rust binding and everything should probably we redone, instead of just changing the dependency for rust.

**Comment by tek** [comment]
I deleted the Rust files and ran `tree-sitter generate`. Didn't change much, but for some reason it referred to `tree_sitter_javascript::language()` before??
I also uncommented the block that recompiles `scanner.c` when changed.

I'm not quite sure I understand your problem description, but since the bindings aren't precompiled, shouldn't everything depend solely on the versions involved when you're building your program?

**Comment by VixieTSQ** [comment]
@tek Okay so with some further exploration into how tree-sitter works, here's a better explanation... here in `lib.rs`
```
extern "C" {
    fn tree_sitter_haskell() -> Language;
}
```
this defines a function that comes from the file `src/parser.c` with a return type of `Language`. Specifically the `Language` struct from the rust tree-sitter crate. My worry is that `parser.c` is outdated and returns a `Language` (the C struct) that is actually different from `Language` (the rust struct). Since the rustc compiler has no knowledge of C and does no checks for this sort of thing, only assuming these two structs are actually the same thing, this would cause UB if there was a breaking change between version 0.19 and version 0.20 of tree-sitter.

**Comment by tek** [comment]
Isn't the parser compiled against a tree-sitter version that's under your control? Or do you depend only on this `Cargo.toml` and the dependency in there is separate from the Rust API that your program uses?

**Comment by VixieTSQ** [comment]
> Isn't the parser compiled against a tree-sitter version that's under your control? Or do you depend only on this `Cargo.toml` and the dependency in there is separate from the Rust API that your program uses?

Err.. I don't understand (maybe I'm just speaking nonsense and wasting your time, who knows)
`parser.c` is just a file,  it's not brought down as a dependency in any dynamic way. Are you saying that `parser.c` actually depends back on rust code at some point? If so, I didn't know. I don't think that makes it a non-problem even so?
My code, using your crate, has the dependency of `tree-sitter = "0.20"`

**Comment by tek** [comment]
do you maybe have an example project to illustrate how this works?

**Comment by VixieTSQ** [comment]
> do you maybe have an example project to illustrate how this works?

You wouldn't need a lot to see issues. Something as simple as this would be UB if there was a breaking change from tree-sitter 0.19 and tree-sitter 0.20

```rust
// Depend on these in cargo.toml
use tree-sitter-haskell::language;
use tree-sitter; // version 0.20

fn main() {
    let lang = language(); // Actually an FFI call to C
    // This used to be a type mismatch,
    // but now tree-sitter-haskell assumes what it gets back from C is a version 0.20 `Language`
    let lang2: tree-sitter::Language = lang;
}
```

**Comment by tek** [comment]
So if we were to synchronize the `version` field in `Cargo.toml` with the version in
```
[dependencies]
tree-sitter = "0.20"
```
this should be transparent, right?

**Comment by VixieTSQ** [comment]
> I assume the Rust bindings aren't published to some package registry, so you're depending on the repo directly? Not sure what the right approach would be to allow you to select a version except for pinning a commit sha.

It is from the repo directly, and I'm just using version 0.14 of your crate as that's what your cargo.toml specifies. It seems to work perfectly fine.

> So if we were to synchronize the `version` field in `Cargo.toml` with the version in
> 
> ```
> [dependencies]
> tree-sitter = "0.20"
> ```
> 
> this should be transparent, right?

That is what we have now. Both tree-sitter-haskell and my project depend on tree-sitter 0.20 and so there is no compiler error. This does not fix the issue whatsoever with the `src/parser.c` which your project makes an FFI call to. It's completely unrelated. `src/parser.c` is still in version 0.19 of tree-sitter, and rust dependency management doesn't do anything with that file and has no knowledge of it.

**Comment by VixieTSQ** [comment]
> In what way is `parser.c` bound to version 0.19? does it use an API feature that was broken in 0.20?

I don't have specifics, I only assume it's different. `parser.c` creates an instance of `Language` (or *something* creates that. All I know is that `parser.c` returns it to rust.) and the definition of the `Language` might be different than it once was in the previous version of tree-sitter. Rust cannot check if it's different or not because rust isn't C.

**Comment by tek** [comment]
Well if you're compiling `parser.c` as part of your project, it will use the `Language` that's present in the C library that your project depends on. If there is an API discrepancy, it will cause a compile error, but there's no ABI involved.

Do you think I'm missing something? Do you see another version of the tree-sitter lib being pulled in somehow?

**Comment by tek** [comment]
oh, right, now I get it!
There appears to be a `struct TSLanguage` defined in `src/tree_sitter/parser.h`. Maybe that's the FFI adapter for Rust's `Language`?
In any case, I don't know what the protocol is for ensuring version consistency here – optimally I guess you'd run `tree-sitter generate` in your project.

**Comment by VixieTSQ** [comment]
`tree-sitter generate` for a binary crate? I'm unsure what the command does exactly. I thought you'd have to look for where ever you pulled all those C files from and get the latest versions of them. Question: what part of this repository did you actually code and isn't just generated? I probably should know this...

**Comment by tek** [comment]
The javascript files and `scanner.c` are written manually, everything else is generated. I think it is recommended practice to not commit the generated files, maybe that would force them to be generated for the right version when depended upon?

I think we need some more competent advice, @maxbrunsfeld

**Comment by VixieTSQ** [comment]
> I think it is recommended practice to not commit the generated files, maybe that would force them to be generated for the right version when depended upon?

No that doesn't sound correct. All the other tree-sitter-* repos commit the generated files. At least in rust, you are never expected to hold a dependency locally unless you wrote it.

**Comment by tek** [comment]
Sounds like this issue isn't really specific to this repo!

**Comment by VixieTSQ** [comment]
> Sounds like this issue isn't really specific to this repo!

either way, couldn't the `parser.c` file just be regenerated now? or why don't you just delete *everything* you didn't make and regenerate it all. That should bring it all up to date, right? Patches shouldn't have breaking changes so you won't have to regenerate for awhile. It's *probably* fine, but definitely annoying.

**Comment by tek** [comment]
The current parser has been generated by tree-sitter 0.20.6. The version in `package.json` does not have an effect on that, only what version my system'd tree-sitter executable is.

**Comment by tek** [comment]
glad we could settle this! :smile:

**Pull request merged**